### PR TITLE
Improve keyword density assessment

### LINKED ISSUE #1309: Improve keyword density assessment
### Description
- The upper boundary should be linked to combination length; it should be higher for more common (shorter) combinations and lower for less common (longer) combinations.
- The upper boundary for the keyword density is raised from 2.5% to 3% (`more than maximum`) and from 3.5% to 4% (`way more than maximum`), respectively.

### Formula

`(keyphrase_frequency / word_count) * 100 * (0.7 + keyphrase_length / 3)`

- `keyphrase_frequency / word_count * 100` is the base formula we've bee using so far. It expresses the keyphrase density in a percentage.

- To this formula a weight was added. The weight consists of the base (0.7), to which a specific weight for the keyphrase length (1/3) is added. These values were obtained by looking at sample texts and reverse-engineering a formula that yields  results which reflect a natural keyword density

### Boundaries
- `way more than maximum`: density > 4%
- `more than maximum`: 3% < density ≤ 4%
- `good`: 0.5% < density ≤ 3%
- `under minimum`: 0% < density ≤ 0.5% _or_ if they keyphrase is used only once in the text

*all of these numbers refer to rounded keyword density

### Feedback
- `way more than maximum`: They keyphrase was found %1$d% times, which is way more than the advised maximum of %2$d% times for a text of this length.
- `more than maximum`: They keyphrase was found %1$d% times, which is more than the advised maximum of %2$d% times for a text of this length.
- `good`: The keyphrase was found %1$d% times, which is great for a text of this length.
- `under minimum`: They keyphrase was found %1$d% times, which is under the recommended minimum of %2$d% times for a text of this length.

### Miscellaneous
- The user should be able to mark the keyword in the text.
- In order to make the feedback not overly complicated, the keyword density percentages should be dropped. Instead, we will provide feedback about the number of times the feedback has been used and how many times it should be used.

### PR DESCRIPTION
**Pull request by manuelaugustin**
## Summary

This PR can be summarized in the following changelog entry:
* Improves the keyword density assessment by taking into account the length of the keyphrase.

## Relevant technical choices:

* An new formula for calculating the keyword density was implemented. This formula takes into account the keyphrase length as a factor.
* The upper boundaries were raised by 0.5% (from 2.5% and 3.5% to 3% and 4%, respectively). The lower boundary remains at 0.5%.
* For low keyword counts, two hard rules have been set. This was done in order encourage users to include keywords also in short texts where the density would otherwise be too high:
  * Regardless of the density, using a keyword only once will result in a bad score and a feedback string encouraging the user to include the keyword more often.
  * Regardless of the density, using the keyword twice will result in a good score.
* The feedbacks strings do not include density percentages anymore. Instead, the maximum/minimum recommended keyword density is provided.  

## Test instructions

This PR can be tested by following these steps:

**Browserified**

* Check out the branch.
* Use the browserified example. Don't forget to run a `grunt build:js`.
* Add a text and a keyword. Makes sure the text has at least 100 words.
* Try out different text lengths, including different numbers of keywords, and varying the length of the keyphrase. Make sure the plugin provides meaningful feedback strings as specified above.

**WordPress**
* Check out the branch `feature/recalibration` in `wordpress-seo`.
* Run yarn install in the plugin's directory.
* Run `yarn add "git+https://github.com/Yoast/yoastseo.js#stories/1309-improve-keyword-density-assessment"` in the plugin's directory.
* Navigate to `node_modules/yoastseo` and run `yarn install` and `grunt build:js`.
* Navigate back to the plugin's directory and run `grunt build:js`.
* Make a post with a text of at least 100 words and a keyword.
* Try out different text lengths, including different numbers of keywords, and varying the length of the keyphrase. Make sure the plugin provides meaningful feedback strings as specified above.

Fixes #1309

### CODE REVIEW

File: src/assessmentHelpers/recommendedKeywordCount.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * @param {number} recommendedKeywordDensity The recommended keyword density (either maximum or minimum).
+ * @param {string} maxOrMin Whether it's a maximum or minimum recommended keyword density.
+ *
+ * @returns {number} The recommended keyword count.
+ */
+module.exports = function( paper, recommendedKeywordDensity, maxOrMin ) {
+	const wordCount = countWords( paper.getText() );
+	const lengthKeyphrase = keyphraseLength( paper );
+
+	let recommendedKeywordCount = ( recommendedKeywordDensity * wordCount ) / ( 100 * ( 0.7 + lengthKeyphrase / 3 ) );
```
Comment (andizer) [inline src/assessmentHelpers/recommendedKeywordCount.js]: I would move this below the `if ( wordCount === 0 ) {`

File: src/assessments/seo/keywordDensityAssessment.js [hunk 2]
```diff
@@ -1,104 +1,201 @@
...
+
+		assessmentResult.setScore( this.calculateScore() );
+		assessmentResult.setText( this.translateScore( i18n ) );
+
+		return assessmentResult;
 	}
 
-	if ( inRangeStartEndInclusive( roundedKeywordDensity, 0.5, 2.5 ) ) {
-		score = 9;
+	/*
```
Comment (andizer) [inline src/assessments/seo/keywordDensityAssessment.js]: Missing `*`. This docbloc should start with `/**`

File: src/assessments/seo/keywordDensityAssessment.js [hunk 3]
```diff
@@ -1,104 +1,201 @@
...
+	 * @returns {boolean} Returns true if the keyword count is 0.
+	 */
+	hasNoMatches() {
+		return this._keywordCount === 0;
+	}
 
-		/* Translators: %1$s expands to the keyword density percentage, %2$d expands to the keyword count. */
-		text = i18n.dgettext( "js-text-analysis", "The keyword density is %1$s, which is great;" +
-			" the focus keyword was found %2$d times." );
+	/*
```
Comment (andizer) [inline src/assessments/seo/keywordDensityAssessment.js]: `/**`

File: src/assessments/seo/keywordDensityAssessment.js [hunk 4]
```diff
@@ -1,104 +1,201 @@
...
+	 * @returns {boolean} Returns true if the rounded keyword density is between 0 and the recommended minimum
+	 * or if there there is only 1 keyword match (regardless of the density).
+	 */
+	hasTooFewMatches() {
+		return inRangeStartInclusive( this._keywordDensity, 0, this._config.minimum ) ||
+			this._keywordCount === 1;
+	}
 
-		text = i18n.sprintf( text, keywordDensityPercentage, keywordCount );
+	/*
```
Comment (andizer) [inline src/assessments/seo/keywordDensityAssessment.js]: `/**`

File: src/assessments/seo/keywordDensityAssessment.js [hunk 5]
```diff
@@ -1,104 +1,201 @@
...
+	 * and the recommended maximum or if the keyword count is 2 and the recommended minimum is lower than 2.
+	 */
+	hasGoodNumberOfMatches() {
+		return inRangeStartEndInclusive( this._keywordDensity, this._config.minimum, this._config.maximum ) ||
+			( this._keywordCount === 2 && this._minRecommendedKeywordCount <= 2 );
 	}
 
-	if ( inRangeStartInclusive( roundedKeywordDensity, 0, 0.5 ) ) {
-		score = 4;
+	/*
```
Comment (andizer) [inline src/assessments/seo/keywordDensityAssessment.js]: `/**`

File: src/assessments/seo/keywordDensityAssessment.js [hunk 6]
```diff
@@ -1,104 +1,201 @@
...
-		text = i18n.sprintf( text, keywordDensityPercentage, keywordCount );
+	/*
+	 * Checks whether there is a good number of keyword matches in the text. Two keyword matches are always
+	 * considered as good, regardless of the density.
+	 *
+	 * @returns {boolean} Returns true if the rounded keyword density is between the recommended minimum
+	 * and the recommended maximum or if the keyword count is 2 and the recommended minimum is lower than 2.
+	 */
+	hasGoodNumberOfMatches() {
+		return inRangeStartEndInclusive( this._keywordDensity, this._config.minimum, this._config.maximum ) ||
```
Comment (andizer) [inline src/assessments/seo/keywordDensityAssessment.js]: This is a pretty long line to read.

File: src/assessments/seo/keywordDensityAssessment.js [hunk 7]
```diff
@@ -1,104 +1,201 @@
...
+const countWords = require( "../../stringProcessing/countWords.js" );
+const inRange = require( "../../helpers/inRange.js" );
+const recommendedKeywordCount = require( "../../assessmentHelpers/recommendedKeywordCount.js" );
+const merge = require( "lodash/merge" );
+
+const inRangeEndInclusive = inRange.inRangeEndInclusive;
+const inRangeStartInclusive = inRange.inRangeStartInclusive;
+const inRangeStartEndInclusive = inRange.inRangeStartEndInclusive;
+
+class KeywordDensityAssessment extends Assessment {
```
Comment (andizer) [inline src/assessments/seo/keywordDensityAssessment.js]: Can you add documentation for this class?

File: src/researches/getKeywordDensity.js [hunk 8]
```diff
@@ -1,19 +1,29 @@
...
- * Calculates the keyword density .
+ * Calculates the keyword density.
  *
  * @param {object} paper The paper containing keyword and text.
  * @param {object} researcher The researcher.
  * @returns {number} The keyword density.
  */
 module.exports = function( paper, researcher ) {
 	const wordCount = countWords( paper.getText() );
+	const lengthKeyphrase = keyphraseLength( paper );
```
Comment (andizer) [inline src/researches/getKeywordDensity.js]: I would move this below `if ( wordCount === 0 ) {`

File: src/researches/getKeywordDensity.js [hunk 9]
```diff
@@ -1,19 +1,29 @@
...
 	const keywordCount = researcher.getResearch( "keywordCount" );
-	return ( keywordCount / wordCount ) * 100;
+
+	/*
+	 * This formula consists of the basic keyword density formula
+	 * (keywordCount / wordCount) * 100, with an added factor for
+	 * keyphrase length. Keyphrase length gets a base weight of 0.7, plus a
+	 * specific weight for the keyphrase length (lengthKeyphrase / 3).
+	 */
+	return ( keywordCount / wordCount ) * 100 * ( 0.7 + lengthKeyphrase / 3 );
```
Comment (andizer) [inline src/researches/getKeywordDensity.js]: I've seen a part of this formula [before](https://github.com/Yoast/YoastSEO.js/pull/1411/files#diff-1fe2b1c3965158b2d0f682d12d347131R18): `100 * ( 0.7 + lengthKeyphrase / 3 )`. You might consider making extracting this.

**Comment by andizer** [comment]
Cr Done 👍

**Comment by jcomack** [comment]
Acceptance done 👍

**Pull request merged**